### PR TITLE
rustdoc: hide `#[repr(transparent)]` if it isn't part of the public ABI

### LINKED ISSUE #90435: Hide `#[repr(transparent)]` where the field is non-public
Per @dtolnay [in #72841](https://github.com/rust-lang/rust/pull/72841#pullrequestreview-429527831):

> Note that repr(transparent) is not the same as permission to inspect or manipulate the internal representation of a type. That permission only exists if the field is also exposed as `pub`.

As such, we should consider hiding `#[repr(transparent)]` when the field is not otherwise visible. This should not be the case when `--document-private-items` is passed for reasons that should be obvious.

This will affect some stdlib structs such as [`NonZeroU8`](https://doc.rust-lang.org/stable/std/num/struct.NonZeroU8.html)

@rustbot label +A-docs +C-enhancement +T-rustdoc

### PR DESCRIPTION
**Pull request by fmease**
Fixes #90435.

This hides `#[repr(transparent)]` when the non-1-ZST field the struct is "transparent" over is private.

CC @RalfJung

Tentatively nominating it for the release notes, feel free to remove the nomination.
@rustbot label needs-fcp relnotes A-rustdoc-ui

### CODE REVIEW

**Comment by fmease** [comment]
r? rustdoc

File: src/doc/rustdoc/src/advanced-features.md
Comment (fmease) [inline src/doc/rustdoc/src/advanced-features.md:114]: I wasn't sure where to place this section, open to suggestions

File: src/librustdoc/clean/types.rs
Comment (fmease) [inline src/librustdoc/clean/types.rs:774]: The early return wasn't correct. It resulted in us not rendering some attributes like `#[non_exhaustive]` in cross-crate re-export scenarios. I've added a regression test for this: `tests/rustdoc/inline_cross/attributes.rs`.

File: src/doc/rustdoc/src/advanced-features.md
Comment (fmease) [inline src/doc/rustdoc/src/advanced-features.md:128]: I'm pretty sure the phrasing in this section can be improved upon, feels a bit awkward in some places.

File: src/librustdoc/json/conversions.rs
Comment (fmease) [inline src/librustdoc/json/conversions.rs:45]: Note that my patch don't affect Rustdoc JSON. We still pass through every inert attribute. Let me know if I that needs changing.

**Comment by GuillaumeGomez** [comment]
Changes look good to me overall (didn't go into details yet). It'll need to go through FCP and to have someone from the compiler team to take a look as well. But otherwise, nice work! :+1:

**Comment by fmease** [comment]
Just so you know, where applicable I'm gonna update the `#[cfg_attr(not(doc), repr(transparent))]` in `library/` to `#[repr(transparent)]` in a separate PR once this one has been approved & merged.

**Comment by fmease** [comment]
Rebased onto latest master to include #115334. Now, there's the open issue #115922 / open PR #115924.
However, they are not blocking this PR, it's ready for review.

**Review by notriddle — APPROVED** [review]

**Comment by Manishearth** [comment]
(in general please make sure to include a short description of the thing that is actually being FCPd since stuff can change through discussion. Fortunately this has in-prose documentation in the PR so it's easy enough to look through)

**Comment by GuillaumeGomez** [comment]
Thanks everyone!

@bors r+ rollup

**Comment by dtolnay** [comment]
Thank you!

I created https://github.com/rust-lang/rust/pull/116743 to follow up by deleting `cfg_attr(not(doc), repr(...))`-based workarounds from the standard library.

**Pull request merged**
